classes/03class/docker/README.md

```diff
@@ -0,0 +1,305 @@
+- an image with a specific version of mysql to run multiple databases
+
+It can also be used to store the final version of your application(artifact) that will run in production
+- an image with apache+php+modules+yourapp
+- an image with python binaries and your application to run a machine learning algorythm
+- an image with terraform cli that executes any terraform code
+
+A Docker registry is often used as a tool in the middle of the CI/CD process, since the Docker image created during the CI pipeline needs to be stored somewhere so it can be used during the CD pipeline. More related to this in class #6.
+
+## Dockerfile
+A Dockerfile is a set of instruction(like a recipe) to create a new Docker image. Those instructions will be used to install the application pre-requisites as well as inlcude all the application related files into the image that is being created.
+
+A Dockerfile is always based on an existen image and the instructions included in the Docker file goes over that base image.
+Example of Dockerfile:
+```
+FROM golang:1.14rc1-buster
+COPY bin/ /
+RUN chmod 544 /app/app1.bin && \
+    chmod 500 /var/logs/* 
+CMD [ "/app/app1.bin", "run" ]
+```
+### Main Dockerfile Instructions
+- FROM
+  - The base image that will be used to build the new image
+- COPY
+  - Copy files between the local filesystem to the container file system
+- ADD
+  - Similar to the COPY instruction, but it can be used to copy directly froma remote URL(http or https)
+- RUN 
+  - Executes a command into the container. The binary that will be executed must reside inside the container
+- ENV
+  - Creates an Environment Variable in the container
+- EXPOSE
+  - Expose a TCP or UDP port to inform Docker that the container will listen on the specified ports
+- USER
+  - Specify which user account will be used to run the commands inside the container when it runs
+- CMD
+  - Set a default command and default parameters which will be executed when Docker is run. Those values can be overwritten during the contianer execution
+- ENTRYPOINT
+  - Used when the container needs to be run as an executable. The commandd inside ENTRYPOINT instruction will can't be overwritten during container execution.
+- WORKDIR
+  - Sets the working directory to run the instructions: RUN, CMD, ENTRYPOINT, COPY and ADD
+
+Reference to all Dockerfile Instructions: [https://docs.docker.com/engine/reference/builder/](https://docs.docker.com/engine/reference/builder/)
+
+### Let's give it a go
+Simple [Docker file](artifacts/Dockerfile):
+
+```bash
+docker build -t website:01 .
+```
+```bash
+docker run -it -p 8081:80 website:01
+```
+
+## Docker Image
+An Image is an ordered collection of root filesystem changes and the corresponding execution parameters for use within a container runtime. An image typically contains an union of layered filesystems stacked on top of each other. An image does not have state and it never changes.
```

Comment:
I'd grab this first sentence and include in the `What is Docker?` section. Since other prior sections were already referring to `images`, we need to introduce it first

Reply:
created a link to the image section